src/app/libp2p_helper/src/codanet.go

```diff
@@ -495,11 +496,15 @@ func (h *Helper) handlePxStreams(s network.Stream) {
 
 		for _, peer := range peers {
 			go func() {
-				err = h.Host.Connect(h.Ctx, peer)
-				if err != nil {
-					logger.Errorf("failed to connect to peer err=%s", err)
+				connInfo := h.ConnectionManager.GetInfo()
+				h.Host.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.ConnectedAddrTTL)
+				if connInfo.ConnCount < connInfo.LowWater {
+					err = h.Host.Connect(h.Ctx, peer)
+					if err != nil {
+						logger.Errorf("failed to connect to peer err=%s", err)
+					}
+					logger.Debugf("connected to peer! %s", peer)
```

Comment:
This log should be in the `else` case (rn, we will print both "failed to connect" and "connected" if we error out.